cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/TransformExecutor.java

```diff
@@ -16,92 +16,125 @@
 
 package co.cask.cdap.etl.common;
 
+import co.cask.cdap.api.metrics.Metrics;
 import co.cask.cdap.etl.api.Destroyable;
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.InvalidEntry;
 import co.cask.cdap.etl.api.Transformation;
-import com.google.common.collect.Lists;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
 
-import java.util.ArrayList;
-import java.util.Collection;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
+
 /**
  * Executes Transforms one iteration at a time, tracking how many records were input into and output from
  * each transform.
  *
  * @param <IN> the type of input object to the first transform
- * @param <OUT> the type of object output by the last transform
+ *
  */
-public class TransformExecutor<IN, OUT> implements Destroyable {
-
-
-  private static final Logger LOG = LoggerFactory.getLogger(TransformExecutor.class);
-
-  private final List<TransformDetail> transforms;
-  private final List<DefaultEmitter> emitters;
+public class TransformExecutor<IN> implements Destroyable {
+
+  private final String start;
+  private final Map<String, List<String>> connectionsMap;
+  private final Map<String, Transformation> trackedTransformation;
+  private final DefaultEmitter<Object> defaultEmitter;
+
+  public TransformExecutor(Map<String, Transformation> transformationMap, Metrics metrics,
```

Comment:
If somebody is only looking at this class, it's confusing what is in transformationMap and what is in connectionsMap.  It's not obvious that something that is in transformationMap but isn't in connectionsMap is a terminal node, and that start could be in the connectionsMap but not in the transformationMap. The caller also needs to make sure that everything else is in both maps.

I think it is more natural to repurpose the old TransformDetail object and pass in a `Map<String, TransformDetail>`, where the detail object contains the actual Transformation, the emitter for that Transformation, and the list of next stages (which may be empty).  TransformDetail could take the stage name, Transformation, Metrics, and connections in its constructor.

That way you wouldn't need to do lookups on connectionsMap and trackedTransformations and the new DefaultEmitter since they would all be together in the same object.